### PR TITLE
Plfm 5841

### PR DESCRIPTION
**Pull request by brucehoff**
In Synapse, HTTP Filters take care of authentication, then pass authenticated identity "down the filter chain" to the controller.   Any information has to be passed somewhere in the incoming request.  This PR makes the following changes:
- OAuthAccessTokenFilter:  Instead of passing the validated token as a request parameter, it is passed along as a request header. Among other benefits, this cleans up the API docs which today show a spurious request param.  (See  https://rest-docs.synapse.org/rest/GET/oauth2/userinfo.html)
- OAuthClientAuthFilter:  After validating the client ID & secret the client ID is now passed as a header rather than as a request parameter. This cleans up the API doc https://rest-docs.synapse.org/rest/POST/oauth2/token.html which today shows a spurious request param for the client ID
- ControllerUtils.processParameterAnnotations(): Currently identifies the reserved request parameter "userId" and switches on "Authentication: required" in the Java doc.  This is changed so it also looks for a request *header* called "Authorization" and, if found, switches on "Authentication: required" in the Java doc.  (Currently we see "Authentication: false" for services that use Access tokens, e.g. https://rest-docs.synapse.org/rest/GET/oauth2/userinfo.html) 
- The AuthenticationController currently mentions three kinds of authentication but doesn't actually say how to implement a client that uses them.  I filled in the missing doc's and added a fourth kind of authentication, i.e. using an access token.

### CODE REVIEW

File: services/authutil/src/test/java/org/sagebionetworks/authutil/ModHttpServletRequestTest.java [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+		assertEquals(-1L, modifiedRequest.getDateHeader(dateHeaderName));
+		
+		// what if we DO set a value?
+		when(originalRequest.getDateHeader(dateHeaderName)).thenReturn(originalDateValue);
+		// method under test
+		assertEquals(originalDateValue, modifiedRequest.getDateHeader(dateHeaderName));
+		
+		// now override the headers
+		Map<String,String[]> headers = new HashMap<String,String[]>();
+		modifiedRequest = new ModHttpServletRequest(modifiedRequest, headers, null);
```
Comment (brucehoff) [inline services/authutil/src/test/java/org/sagebionetworks/authutil/ModHttpServletRequestTest.java:53]: originalRequest not modifiedRequest

File: services/authutil/src/test/java/org/sagebionetworks/authutil/ModHttpServletRequestTest.java [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+		ModHttpServletRequest modifiedRequest = new ModHttpServletRequest(originalRequest, null, null);
+		when(originalRequest.getDateHeader(dateHeaderName)).thenReturn(-1L);
+		// what if we don't set any value?
+		// method under test
+		assertEquals(-1L, modifiedRequest.getDateHeader(dateHeaderName));
+		
+		// what if we DO set a value?
+		when(originalRequest.getDateHeader(dateHeaderName)).thenReturn(originalDateValue);
+		// method under test
+		assertEquals(originalDateValue, modifiedRequest.getDateHeader(dateHeaderName));
```
Comment (brucehoff) [inline services/authutil/src/test/java/org/sagebionetworks/authutil/ModHttpServletRequestTest.java:49]: verify(originalRequest).getDateHeader()

File: services/authutil/src/test/java/org/sagebionetworks/authutil/ModHttpServletRequestTest.java [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+		
+		// now override the headers
+		Map<String,String[]> headers = new HashMap<String,String[]>();
+		modifiedRequest = new ModHttpServletRequest(modifiedRequest, headers, null);
+		
+		// check the case of no such header in the overriding list
+		// method under test
+		assertEquals(-1L, modifiedRequest.getDateHeader(dateHeaderName));
+		// what if we DO set a value?
```
Comment (brucehoff) [inline services/authutil/src/test/java/org/sagebionetworks/authutil/ModHttpServletRequestTest.java:58]: verify(originalRequest, never()).getDateHeader();

File: services/authutil/src/test/java/org/sagebionetworks/authutil/ModHttpServletRequestTest.java [hunk 4]
```diff
@@ -0,0 +1,346 @@
...
+		assertNull(modifiedRequest.getHeader(headerName));
+		
+		// what if we DO set a value?
+		when(originalRequest.getHeader(headerName)).thenReturn(originalValue);
+		// method under test
+		assertEquals(originalValue, modifiedRequest.getHeader(headerName));
+		
+		// now override the headers
+		Map<String,String[]> headers = new HashMap<String,String[]>();
+		modifiedRequest = new ModHttpServletRequest(modifiedRequest, headers, null);
```
Comment (brucehoff) [inline services/authutil/src/test/java/org/sagebionetworks/authutil/ModHttpServletRequestTest.java:86]: originalRequest

File: services/authutil/src/test/java/org/sagebionetworks/authutil/ModHttpServletRequestTest.java [hunk 5]
```diff
@@ -0,0 +1,346 @@
...
+		
+		// check the case of no such header in the overriding list
+		// method under test
+		assertNull(modifiedRequest.getHeader(headerName));
+		// what if we DO set a value?
+		headers.put(headerName, new String[] {overrideValue, "some other value"});
+		modifiedRequest = new ModHttpServletRequest(modifiedRequest, headers, null);
```
Comment (brucehoff) [inline services/authutil/src/test/java/org/sagebionetworks/authutil/ModHttpServletRequestTest.java:93]: originalRequest

File: services/authutil/src/test/java/org/sagebionetworks/authutil/ModHttpServletRequestTest.java [hunk 6]
```diff
@@ -0,0 +1,346 @@
...
+		assertFalse(modifiedRequest.getHeaders(headerName).hasMoreElements());
+		
+		// what if we DO set a value?
+		when(originalRequest.getHeaders(headerName)).thenReturn(Collections.enumeration(Arrays.asList(originalValues)));
+		// method under test
+		assertTrue(enumerationsAreEqual(Collections.enumeration(Arrays.asList(originalValues)), modifiedRequest.getHeaders(headerName)));
+		
+		// now override the headers
+		Map<String,String[]> headers = new HashMap<String,String[]>();
+		modifiedRequest = new ModHttpServletRequest(modifiedRequest, headers, null);
```
Comment (brucehoff) [inline services/authutil/src/test/java/org/sagebionetworks/authutil/ModHttpServletRequestTest.java:140]: originalRequest

File: services/authutil/src/test/java/org/sagebionetworks/authutil/ModHttpServletRequestTest.java [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+import com.google.common.collect.ImmutableMap; 
+
+@ExtendWith(MockitoExtension.class)
+public class ModHttpServletRequestTest {
+	
+	
+	@Mock
+	private HttpServletRequest originalRequest;
+	
+	@Test
```
Comment (brucehoff) [inline services/authutil/src/test/java/org/sagebionetworks/authutil/ModHttpServletRequestTest.java:33]: split all tests up

File: services/repository/src/main/java/org/sagebionetworks/auth/HttpAuthUtil.java [hunk 8]
```diff
@@ -24,10 +31,50 @@ public static UserNameAndPassword getBasicAuthenticationCredentials(HttpServletR
 	}
 
 	public static String getBearerToken(HttpServletRequest httpRequest) {
-		
 		String header = httpRequest.getHeader(AuthorizationConstants.AUTHORIZATION_HEADER_NAME);
+		return getBearerTokenFromAuthorizationHeader(header);
+	}
+
+	public static String getBearerTokenFromAuthorizationHeader(String header) {
```
Comment (brucehoff) [inline services/repository/src/main/java/org/sagebionetworks/auth/HttpAuthUtil.java:38]: check StringUtils.isBlank()
trim result

File: services/repository/src/main/java/org/sagebionetworks/auth/HttpAuthUtil.java [hunk 9]
```diff
@@ -24,10 +31,50 @@ public static UserNameAndPassword getBasicAuthenticationCredentials(HttpServletR
...
+			String headerName = e.nextElement();
+			if (AUTHORIZATION_HEADERS_LOWER_CASE.contains(headerName.toLowerCase())) {
+				continue;
+			}
+			List<String> headerValues = new ArrayList<String>();
+			for (Enumeration<String> n = request.getHeaders(headerName); n.hasMoreElements();) {
+				String headerValue = n.nextElement();
+				headerValues.add(headerValue);
+			}
+			result.put(headerName, headerValues.toArray(new String[] {}));
```
Comment (brucehoff) [inline services/repository/src/main/java/org/sagebionetworks/auth/HttpAuthUtil.java:76]: new String[headerValues.size()]

File: services/repository/src/main/java/org/sagebionetworks/auth/OpenIDConnectController.java [hunk 10]
```diff
@@ -252,57 +256,53 @@ OAuthAuthorizationResponse authorizeClient(
...
 			@RequestParam(value = AuthorizationConstants.OAUTH2_SCOPE_PARAM, required=false) String scope,
 			@RequestParam(value = AuthorizationConstants.OAUTH2_CLAIMS_PARAM, required=false) String claims,
 			UriComponentsBuilder uriComponentsBuilder
 			)  throws NotFoundException {
-		if (StringUtils.isEmpty(userId)) {
-			// the 'userId' param is actually the OAuth client ID, verified by Basic Authentication
-			throw new UnauthenticatedException("OAuth Client ID and secret must be passed via Basic Authentication.  Credentials are missing or invalid.");
-		}
-		return serviceProvider.getOpenIDConnectService().getTokenResponse(userId, grant_type, code, redirectUri, refresh_token, scope, claims, getEndpoint(uriComponentsBuilder));
+		return serviceProvider.getOpenIDConnectService().getTokenResponse(verifiedClientId, grant_type, code, redirectUri, refresh_token, scope, claims, getEndpoint(uriComponentsBuilder));
```
Comment (brucehoff) [inline services/repository/src/main/java/org/sagebionetworks/auth/OpenIDConnectController.java]: return 401

File: services/repository/src/test/java/org/sagebionetworks/auth/HttpAuthUtilTest.java [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+
+	@Test
+	void testGetBearerToken() {
+		// test no authorization header
+		when(httpRequest.getHeader("Authorization")).thenReturn(null);
+		// method under test
+		assertNull(HttpAuthUtil.getBearerToken(httpRequest));
+
+		// proper bearer token
+		when(httpRequest.getHeader("Authorization")).thenReturn("Bearer "+BEARER_TOKEN);
```
Comment (brucehoff) [inline services/repository/src/test/java/org/sagebionetworks/auth/HttpAuthUtilTest.java:66]: more white space

File: services/repository/src/test/java/org/sagebionetworks/auth/HttpAuthUtilTest.java [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+		// method under test
+		assertEquals(BEARER_TOKEN, HttpAuthUtil.getBearerToken(httpRequest));
+}
+
+	@Test
+	void testGetBearerTokenFromAuthorizationHeader() {
+		// method under test
+		assertNull(HttpAuthUtil.getBearerTokenFromAuthorizationHeader(null));
+		// method under test
+		assertNull(HttpAuthUtil.getBearerTokenFromAuthorizationHeader("Basic xxx"));
```
Comment (brucehoff) [inline services/repository/src/test/java/org/sagebionetworks/auth/HttpAuthUtilTest.java:76]: add white space to check trim, etc.

**Review by marcomarasca — APPROVED** [review]

**Pull request merged**